### ISSUE TITLE
Party ID handling is a mess

### ISSUE THREAD

**Issue by cocreature** [digital-asset/daml#7304]
We have more and more ledgers where you cannot choose parties freely and need to rely on whatever party id you get back from the party management service. We do have a party id hint field with no specified behavior. There are currently two ways in which this is handled by ledgers:

1. Used directly as the party, failing if the party id has already been allocated. This gets you back to a state where you can rely on fixed party ids.
2. Always return the fresh party, the hint is either ignored completely or part of the resulting ID but never the full party id.

Most ledgers that do 2 have to do this in a production setup so I think rather than trying to encourage relying on fixed party ids and forcing ledgers to implement 1 at least behind some flag, we should fix the client side to handle this more nicely.

This is a list of the issues and inconsistencies that I’m currently aware off:

1. Navigator: By default it doesn’t use the party management API at all. `daml ledger navigator` works around this by querying the party management API once before starting navigator and generating a config based on this mapping display name to party id. However, that still doesn’t help you if you want to create a template, choice, …. The form will expect you to fill in a party.

2. JWTs on the Ledger and the JSON API in particular: You have to specify the actual party id in there so if this changes after every restart you end up having to regenerate your tokens constantly which really only works if you can generate them automatically and we have no tooling for that in the SDK.

3. `daml ledger`: Always sets identical display name and party id hint when allocating parties, relies on uniqueness of display name which is not enforced on any ledger atm.

4. DAML Script API: Makes it hard to hardcode party ids, it makes it easy to allocate a party with a given display name but a (tiny) bit harder to allocate one with a given hint.

5. `daml test-script`: Relies on tests being isolated by allocating fresh parties with a given display name. This only works if no uniqueness on display names is not enforced.

6. DAML-LF: Only shows party id (and has only access to that). It would be great to get the display name for debugging purposes.

**Comment by bame-da** [comment]
I would like to solve this around the mental model that there is

- a global identifier (the party Id) and
- a local identifier (currently the display name). 

The global identifier tends to have cryptographic meaning on the ledger or may be entirely random so it's not nice to use for client apps. Its place is below the Ledger API. 
The local identifier has meaning outside of the ledger, eg by identifying a user. When interacting with a ledger via the Ledger API, I typically want to use the local identifier over which I have more control.

It's important that both are unique. I would therefore propose the following:

When allocating a party, I send a display name (ie deprecate hint). The participant tries to generate

1. Entropy for the local name. The local identifier is `"display name:entropy")`
2. A global identifier that is as close as possible to the local identifier (ie the local identifier is the id hint). But the hint is always just a hint. Ie the hint should never be a reason a party allocation request is rejected.

The `Party` API type stays as is, representing the global ID, but the party allocation and query services are extended/changed:

1. Party allocation takes the display name only by default. You can set a custom hint for backwards compatibility, but we can always ignore that for all I care.
2. Party service allows you to query by display name, local identifier, or global name, and always get all three for all matching parties back.

Now all our tooling should allow working with display name or local identifier. Numbers matching 1 - 6 from the issue description:

1. Navigator shows local identifier on login page, shows local and global identifier on contracts, and allows specifying Parties by local or global identifier - I'd imagine a dropdown with autocomplete that matches on either.
2. JWT claims can contain display name or local identifier. If they contain display name, and the display name is not unique, the request fails.
3. `daml ledger` sets display name, as it does now, but doesn't deduplicate. It returns local and global identifier.
4. DAML Script behaves the same as `daml ledger`. ie `allocateParty` in Script is equivalent to `daml ledger allocate-parties`.
5. `daml test-script` still works as it does now
6. This is the one thing I'm unsure about. Querying the list of parties is currently Admin functionality. Is one user allowed to know the display name of another users' party? Assuming yes, what do we do with remote parties? Do we generate a local identifier for newly learnt parties parties?  Does the Participant Admin get to set local identifiers for remote parties? If your UI is meant to show nice names, I'd much rather you made that an application concern than a Ledger concern, I think.

**Comment by bame-da** [comment]
Actually, introducing extra entropy is pointless. local identifier could just be `display name:global identifier`.

**Comment by cocreature** [comment]
Not really sure how the difference between local and global IDs addesses the UX issues here. Is the point that the local name is guaranteed to provide a mapping to the display name whereas the globlal one does not need to preserve this?

Let me try to summarize your proposal to address the UX issues (omitting uninteresting points) pointing to potential issues:

1. Navigator: Autocomplete relying on the fact that the human readable name is included in the party id (at least the local one).
2. JWTs: Allow display names, since you didn’t specify this explicitly this needs to be handled in the ledger APi not the JSON APi otherwise it doesn’t work if auth is enabled. It feels a bit wrong to allow display names in a production setting and I’m slightly worried about potential security implications (I don’t have a concrete issue but it seems fishy). Maybe this should be hidden behind a flag?
6. Let’s say I have `assert (controller `elem` …)` which is something you have fairly often in combination with flexible controllers. To do the translation on failures at the app-level I either have to parse the string and replace the party ids (:vomiting_face:) or reimplement the logic for those assertions on the client side (also :vomiting_face:). You do raise a good point, that this is currently admin-only information. I’m not sure it has to be but I’ll think about it a bit more.

**Comment by bame-da** [comment]
If we say that display names are public information, and all DAML Ledgers allow prefixing of party IDs, we could follow @garyverhaegen-da 's idea (and current behaviour in Canton) and just guarantee that global identifier = "displayName" <> someReservedSymbol <> "LedgerSpecificEntropy". We could then follow GIT's lead on references, and on the write pipeline allow any initial segment of the global identifier that contains at least the displayName, as long as that initial segment is unique.

**Comment by cocreature** [comment]
If we do that, what purpose does the distinction between global and local identifier serve?

**Comment by robin-da** [comment]
:+1: for "displayName" <> someReservedSymbol <> "LedgerSpecificEntropy" and no local ids. The concept is easy to understand, and it allows autocomplete in an UI.

**Comment by daravep** [comment]
So, Canton uses `alice::01c8c7795ef6c984ef056fca33983d12f769c64cfd9d24cfa0f080df386ddb3810`, where `alice` is the `party-hint` to define a party. The separator is `::` and the suffix is the `identity namespace key fingerprint`. We also support `participant.parties.set_display_name(partyId, displayName)`, however, display names are kept separate from the party id in order to provide privacy and anonymity. 

I.e. we generally use "display names" for parties in tests and demos. However, for "real systems", we suggest to use a random number for the party identifier to provide pseudonymity. In other words, there is a separation between "legal identity" and "system identity" and I think it is important to keep this. Otherwise, I already see the following scenarios happening: Scanner creating a new party `goldmansachs::01c8c7795ef6c984ef056fca33983d12f769c64cfd9d24cfa0f080df386ddb3810`.

**Comment by maxdeliso-da** [comment]
I have spent some time considering these problems in detail, and I believe there is a lot of intertwined complexity in party identity and party allocation in particular.

I'd like to take a swing at formalizing a lexicon in an effort to craft a semantic workbench atop which clarity may be forged.

Additionally I confess to the pretension that I believe my concepts of these problems are useful in that they are grounded in experience building services that operationalize them successfully.

The different conceptions related to the party id are as such:
* **the authority to act as the party**, which is asserted by a person or system bearing a cryptographically verifiable token with the party literal encoded within it
* **the system identity of the party**, or its "party literal," as mentioned above this one is encoded within access token in a fully authenticated setup
* **a symbolic label for the party**, or a human consumable label. this frequently accompanies the party, sometimes in a JWT as well. it is used by people to understand what the party is for
* **a party hint**, which is an argument to the party management service, which is to be understood as a hint to the allocator which *could* be used by the allocator to include some semantically identifying text within the allocated party literal, could be literally used _as_ the part literal, and also could be completely ignored (hypothetically)

I think the reason the party hint argument to the party management service party allocation API is confusing is because it _conflates the notion of a system identity and a symbolic label_. It may be less confusing to _always_ have only machine readable identifiers as party literals (for instance BTC addresses work this way). Currently the shendlibonx uses the hint _as_ the party literal, which perpetuates the confusion described in detail here.

Certainly it is important for humans to understand the intended use of a party, and that is what a symbolic or human readable label is for. I also agree that it's important to have both concepts. But, to me, the label is incidental metadata, and should be thought of as completely distinct from the authority to act as the party, as well as the system identity of the party.

Another way in which the symbolic and system identities are different is that the system identifiers **must** be globally unique, whereas the symbolic identities need not be. Bootstrapping identity, in my opinion, should never involve examination of party _metadata_, but instead should be built atop the authority attached to the party itself - the creation and exercise of contracts. This is somewhat tangential to the discussion, though.

In summary, I'm of the opinion that party literals should not have any semantic significance at all. I believe having symbolic labels as party literals is a harmful oversimplification that leads to increased difficulty in introducing authentication and authorization related concepts later on.

**Comment by garyverhaegen-da** [comment]
@cocreature @bame-da this discussion predates Daml 2.0. Is it still relevant, or can the ticket be closed?
